File: retrobiocat_web/app/database_bps/curation/routes/ajax/paper_tab_ajax/update_paper_metadata_ajax.py
import json
from distutils.util import strtobool

from flask import request, jsonify
from flask_login import current_user
from flask_security import roles_required

from retrobiocat_web.app.database_bps.curation import bp
from retrobiocat_web.app.database_bps.curation.functions import check_permission
from retrobiocat_web.mongo.modal_updates import paper_CUD
from retrobiocat_web.mongo.model_queries import paper_queries
from retrobiocat_web.mongo.models.user_models import user_datastore


@bp.route('/_save_updated_paper_metadata', methods=['GET', 'POST'])
@roles_required('contributor')
def save_updated_paper_metadata():
    """ Save updates to the metadata for a paper """

    user = user_datastore.get_user(current_user.id)
    paper = paper_queries.paper_from_id(request.form['paper_id'])
    update_dict = json.loads(request.form['update_dict'])

    issues = paper_CUD.update_paper(paper, update_dict, user=user)

    if len(issues) == 0:
        result = {'status': 'success',
                  'msg': 'Paper information updated',
                  'issues': issues}
    else:
        result = {'status': 'danger',
                  'msg': 'Problem updating paper data',
                  'issues': issues}

    return jsonify(result=result)

@bp.route('/_update_paper_importance', methods=['GET', 'POST'])
@roles_required('contributor')
def update_paper_importance():
    paper = paper_queries.paper_from_id(request.form['paper_id'])
    if not check_permission.check_review_persmission(current_user.id, paper):
        result = {'status': 'danger',
                  'msg': 'No access to update importance of this paper',
                  'issues': [f'No paper permission for {paper.short_citation}']}
        return jsonify(result=result)

    importance = bool(strtobool(request.form['importance']))
    issues = paper_CUD.update_paper_importance(paper, importance)

    if len(issues) == 0:
        result = {'status': 'success',
                  'msg': 'Paper importance updated',
                  'issues': []}
    else:
        result = {'status': 'danger',
                  'msg': 'Error updating paper importance',
                  'issues': issues}
    return jsonify(result=result)

File: retrobiocat_web/app/database_bps/curation/static/js/ajax/paper_metadata_ajax/update_paper_metadata_ajax.js
var update_paper_metadata_ajax = (function () {
    /*
    Methods for making ajax calls to update paper metadata fields
    */


    function save_updated_paper_metadata(paper_id, update_dict, callback, response_div, on_finish) {
        $.post('/_save_updated_paper_metadata', {
            paper_id: paper_id,
            update_dict: JSON.stringify(update_dict)
            }).done(function(data) {
                response_msg(data.result.msg, data.result.status, data.result.issues, response_div)
                if (data.result.status === 'success') {
                    callback()
                }
            }).fail(function (xhr, status, error) {
                response_msg('JS error while trying to update paper metadata', 'danger', [error], response_div)
                console.log(error)
            }).always(function () {
                if (typeof on_finish !== 'undefined') {
                    on_finish()
                }
            })
        }

    function update_paper_importance(paper_id, important_bool, response_div) {
        $.post('/_update_paper_importance', {
            paper_id: paper_id,
            importance: important_bool
            }).done(function(data) {
                response_msg(data.result.msg, data.result.status, data.result.issues, response_div)
            }).fail(function (xhr, status, error) {
                response_msg('JS error while trying to update paper metadata', 'danger', [error], response_div)
                console.log(error)
            })
        }


    return {
        save_updated_paper_metadata: save_updated_paper_metadata,
        update_paper_importance: update_paper_importance
    }

})()